### PR TITLE
Use thread names in logs and deadlock diagnostics

### PR DESCRIPTION
**Pull request by jamesob**
While trying to debug the apparent deadlock revealed in https://github.com/bitcoin/bitcoin/pull/12873, I realized it'd be nice to have the `POTENTIAL DEADLOCK DETECTED` message display which thread acquired each lock. I also think it'd be generally useful to have log lines display the name of the originating thread.

This changeset does both of those things by introducing a class which manages thread naming, `ThreadNameRegistry`. The class abstracts process-control calls responsible for thread naming and provides automatic number suffixing to threads which use the same name (e.g. `httpworker.0`, `httpworker.1`, ...).

With this patch, thread names look like this

```
 $ pstree -p `pgrep bitcoind`

bitcoind(3415)─┬─{addcon}(3465)
               ├─{httpworker.0}(3454)
               ├─{httpworker.1}(3455)
               ├─{httpworker.2}(3456)
               ├─{httpworker.3}(3457)
               ├─{http}(3453)
               ├─{msghand}(3467)
               ├─{net}(3463)
               ├─{opencon}(3466)
               ├─{scheduler}(3452)
               ├─{scriptch.0}(3445)
               ├─{scriptch.1}(3446)
               ├─{scriptch.2}(3447)
               ├─{scriptch.3}(3448)
               ├─{scriptch.4}(3449)
               ├─{scriptch.5}(3450)
               ├─{scriptch.6}(3451)
               └─{torcontrol}(3462)
```
and the debug log looks like this
```
2018-04-26T21:54:23Z [bitcoind] init message: Loading wallet...
2018-04-26T21:54:24Z [bitcoind] mapBlockIndex.size() = 1
2018-04-26T21:54:24Z [bitcoind] nBestHeight = 0
2018-04-26T21:54:24Z [loadblk] Imported mempool transactions from disk: 0 succeeded, 0 failed, 0 expired, 0 already there
2018-04-26T21:54:24Z [torcontrol] torcontrol thread start
2018-04-26T21:54:24Z [bitcoind] Bound to [::]:18444
2018-04-26T21:54:24Z [torcontrol] tor: Error connecting to Tor control socket
2018-04-26T21:54:24Z [torcontrol] tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
2018-04-26T21:54:24Z [bitcoind] Bound to 0.0.0.0:18444
2018-04-26T21:54:24Z [bitcoind] init message: Loading P2P addresses...
2018-04-26T21:54:24Z [bitcoind] Loaded 0 addresses from peers.dat  0ms
2018-04-26T21:54:24Z [dnsseed] dnsseed thread start
2018-04-26T21:54:24Z [bitcoind] init message: Done loading
2018-04-26T21:54:24Z [addcon] addcon thread start
2018-04-26T21:54:24Z [dnsseed] Loading addresses from DNS seeds (could take a while)
2018-04-26T21:54:24Z [net] net thread start
2018-04-26T21:54:24Z [dnsseed] 0 addresses found from DNS seeds
2018-04-26T21:54:24Z [dnsseed] dnsseed thread exit
2018-04-26T21:54:24Z [msghand] msghand thread start
2018-04-26T21:54:24Z [opencon] opencon thread start
2018-04-26T21:54:25Z [torcontrol] tor: Error connecting to Tor control socket
2018-04-26T21:54:25Z [torcontrol] tor: Not connected to Tor control port 127.0.0.1:9051, trying to reconnect
```

Note that child thread names have changed; I'm no longer using the `bitcoin-` prefix. Because we're limited to 16 characters for thread name (on Linux, anyway), that prefix was causing the numeric suffix some names have to be hidden. If it's desirable to keep the prefix, I can revert this change.

#### Implementation considerations

A basic version of this change could be done pretty trivially with something like 
```cpp
thread_local std::string threadname;
```
but per @theuni (https://github.com/bitcoin/bitcoin/pull/11722#pullrequestreview-79322658), we can't rely on having `thread_local`. Also note that with a basic implementation like that, we wouldn't be able to do the automatic numbering scheme to differentiate between threads with the same basename (e.g. `httpworker`, `scriptch`).

We could also rely solely on thread-related system calls. I don't like this much either because of [how poorly defined or unavailable `getname` is on some platforms](https://stackoverflow.com/a/7989973), e.g. OS X, FreeBSD; not to mention the 16 character limit.

#### Tests

~~If this gets a Concept ACK or two, I'll implement some.~~ Unittests attached.

### CODE REVIEW

File: src/threadnames.cpp [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+#ifdef HAVE_SYS_PRCTL_H
+#include <sys/prctl.h> // For prctl, PR_SET_NAME
+#endif
+
+
+extern int LogPrintStr(const std::string &str);
+
+std::unique_ptr<ThreadNameRegistry> g_thread_name_registry(new ThreadNameRegistry());
+
+namespace {
```
Comment (jamesob) [inline src/threadnames.cpp]: Is there somewhere more general or useful I should put the two functions in this namespace?

File: src/threadnames.cpp [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+    std::lock_guard<std::mutex> guard(m_map_lock);
+
+    if (expect_multiple) {
+        name = tfm::format("%s.%d", name, CountMapPrefixes(m_name_to_id, name));
+    }
+
+    std::string process_name(name);
+
+    /*
+     * Uncomment if we want to retain the `bitcoin-` system thread name prefix.
```
Comment (jamesob) [inline src/threadnames.cpp]: Remove or uncomment this block before merge.

File: src/threadnames.cpp [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    m_id_to_name[thread_id] = name;
+    m_name_to_id[name] = thread_id;
+    return true;
+}
+
+const std::string& ThreadNameRegistry::GetThreadName()
+{
+    auto thread_id = boost::this_thread::get_id();
```
Comment (promag) [inline src/threadnames.cpp]: Should lock mutex.

File: src/threadnames.cpp [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+    if (found != m_id_to_name.end()) {
+        return found->second;
+    } else {
+        auto insert_ret = m_id_to_name.emplace(thread_id, GetProcessThreadName());
+        return insert_ret.first->second;
+    }
+}
+
+void ThreadNameRegistry::RenameProcessThread(const char* name)
```
Comment (promag) [inline src/threadnames.cpp]: This could be a static function or a function in a anonymous namespace.

File: src/threadnames.cpp [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+#elif defined(MAC_OSX)
+    pthread_setname_np(name);
+#else
+    // Prevent warnings for unused parameters...
+    (void)name;
+#endif
+}
+
+std::string ThreadNameRegistry::GetProcessThreadName()
```
Comment (promag) [inline src/threadnames.cpp]: This could be a static function or a function in a anonymous namespace.

File: src/threadnames.cpp (hunk 4 above)
Comment (promag) [reply src/threadnames.cpp]: Argument could be `const std::string& name`?

File: src/threadnames.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+const std::string& ThreadNameRegistry::GetThreadName()
+{
+    auto thread_id = boost::this_thread::get_id();
+    auto found = m_id_to_name.find(thread_id);
+
+    if (found != m_id_to_name.end()) {
+        return found->second;
+    } else {
+        auto insert_ret = m_id_to_name.emplace(thread_id, GetProcessThreadName());
+        return insert_ret.first->second;
```
Comment (promag) [inline src/threadnames.cpp]: Shoud also add to `m_name_to_id`?

File: src/threadnames.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+#include <boost/thread.hpp>
+
+
+/**
+ * Keeps a map of thread IDs to string names and handles system-level thread naming.
+ */
+class ThreadNameRegistry
+{
+public:
+    const std::string& GetThreadName();
```
Comment (promag) [inline src/threadnames.h]: Why return reference?

**Review by promag — COMMENTED** [review]
Some comments.

File: src/threadnames.cpp (hunk 6 above)
Comment (jamesob) [reply src/threadnames.cpp]: Good catch, thanks.

**Comment by skeees** [comment]
Thoughts on pulling the `TraceThread` class out of util into the new files you're creating here? Seems like a good home for tracethread and would reduce the clutter in util

File: src/threadnames.cpp (hunk 4 above)
Comment (jamesob) [reply src/threadnames.cpp]: I'd like to be able to unittest the `*Process*` functions eventually, so maybe I should just make them public.

**Comment by practicalswift** [comment]
Concept ACK

**Comment by jamesob** [comment]
@promag thanks for the review; I've pushed changes incorporating your feedback.

@skeees I like that idea but maybe we can save it for a future PR? That entails changing a bunch of call sites and I'm worried about introducing a circular dependency between util and threadnames (by way of logging).

**Comment by TheBlueMatt** [comment]
How does the use of boost::thread::id interact with eventual boost::thread removal? IIRC there was something about no guarantees being made that you can interact with boost::thread if the thread was started by std, though it usually works in practice. We do have a HAVE_THREAD_LOCAL defined, which should be set everywhere but OSX, so maybe we could just turn it on there if boost::thread::id is an issue?

File: src/threadnames.h [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * @return the system's name for the current thread.
+     */
+    std::string GetProcessName();
+
+private:
+    std::mutex m_map_lock;
+    std::map<boost::thread::id, std::string> m_id_to_name;
+    std::map<std::string, boost::thread::id> m_name_to_id;
```
Comment (conscott) [inline src/threadnames.h]: Maybe `m_name_to_id` could instead be a `map<string, int>` - where it just stores the occurrences of that name/prefix, then you don't have to call `CountMapPrefix`, but can just update the count each time.

Comment (jamesob) [reply src/threadnames.h]: Thanks for the look and great suggestion. I think maintaining a map of counts will also allow me to remove all of the utility functions in the anonymous namespace - nice catch!

**Comment by jamesob** [comment]
@TheBlueMatt good points, thanks for looking. I've pushed changes which
- only make use of `boost::thread` when `thread_local` is not available. When it is available, we assign each thread a `thread_local`-cached ID based on a static counter.
- incorporate @conscott's suggestion of using `m_name_to_count` which nicely cuts out a bunch of code.
- add tests.

([jamesob/threadnames.1 -> jamesob/threadnames.3](https://gist.github.com/jamesob/88f7cf082ee0bf97a2411bf8bd120594))

This changeset is ready for review-for-merge IMO.

**Comment by sipa** [comment]
@theuni What's the status of `thread_local` on OSX? I remember seeing https://stackoverflow.com/questions/28094794/why-does-apple-clang-disallow-c11-thread-local-when-official-clang-supports, but which Xcode version's patches does our clang run with, or...?

**Comment by jamesob** [comment]
I've pushed a small change removing `GetProcessName` support for *BSD systems. After double-checking `pthread_getname` support for OpenBSD, I found that [it doesn't seem to exist](https://man.openbsd.org/pthreads.3). As a result, any threads which are not explicitly named with `Rename()` will be labeled `"<unknown>"` on BSD systems.

```diff
 $ diff -uw <(git diff master..threadnames.3) <(git diff master..threadnames.4)

--- /proc/self/fd/12    2018-05-01 14:56:59.561313752 -0400
+++ /proc/self/fd/13    2018-05-01 14:56:59.549313568 -0400
@@ -2215,10 +2215,10 @@
      void reset();
 diff --git a/src/threadnames.cpp b/src/threadnames.cpp
 new file mode 100644
-index 0000000..305c3b4
+index 0000000..92e3f75
 --- /dev/null
 +++ b/src/threadnames.cpp
-@@ -0,0 +1,135 @@
+@@ -0,0 +1,133 @@
 +// Copyright (c) 2018 The Bitcoin Core developers
 +// Distributed under the MIT software license, see the accompanying
 +// file COPYING or http://www.opensource.org/licenses/mit-license.php.
@@ -2345,8 +2345,6 @@
 +
 +#if defined(PR_GET_NAME)
 +    ::prctl(PR_GET_NAME, pthreadname_buff);
-+#elif (defined(__FreeBSD__) || defined(__OpenBSD__) || defined(__DragonFly__))
-+    pthread_get_name_np(pthread_self(), pthreadname_buff);
 +#elif defined(MAC_OSX)
 +    pthread_getname_np(pthread_self(), pthreadname_buff, sizeof(threadname_buff));
 +#else
```

**Comment by jamesob** [comment]
After spending the day figuring out that [`thread_local` is apparently broken on mingw32 compilers](https://gist.github.com/jamesob/fe9a872051a88b2025b1aa37bfa98605), I'm now falling back to boost thread IDs for WIN32. Tests are passing and this is once again ready for review.

**Comment by jamesob** [comment]
Per @sipa's advice, I'm now using `std::thread` to determine thread IDs for all platforms instead of either boost or the ham-fisted `thread_local` code.

[jamesob/threadnames.11 -> jamesob/threadnames.13](https://gist.github.com/jamesob/310427d0590b1790b7601afdf86c09c8)

**Comment by theuni** [comment]
If platform-specific functions are being used anyway, why not just use pthread_getspecific/pthread_setspecific, which are meant for exactly this? That also makes it a trivial transition when we're finally ready to use thread_local.

**Comment by theuni** [comment]
@jamesob See the top two commits here for a quick attempt: https://github.com/theuni/bitcoin/commits/thread_names

That addresses a few things that make me hesitate about this PR:
- system thread id/name are conflated with internal use. This drags in system requirements (name length, for ex) and overhead for internal functions.
- Heavy map lookup with locking. #12970 demonstrated how painful overhead in logging can be.
- Most can go away once we can use thread_local unconditionally.

I realize that my changes are oversimplified, it's just a quick POC for discussion.

**Comment by jamesob** [comment]
@theuni I think your simple implementation looks great. Very much agree with your concern re: map lookup and locking overhead, and I think there's a way to combine our approaches to avoid it. Let me know if this sounds right:

IMO lock acquisition during the single `Rename()` call per thread is acceptable, but we want to avoid any subsequent locking during `GetName()` (since it is called per logline). So: during `Rename()` we can use Registry's count map for numeric suffix generation and then use your method for caching the name of the thread for use during `GetName()`.

If that sounds good to you, I'll cherry-pick your two commits onto this branch and reorganize accordingly. Otherwise if you think the numeric suffix isn't worthwhile, I'm happy to close this out and proceed on your commits only.

**Comment by theuni** [comment]
@jamesob Looking again, I think handling the suffix at this layer is unnecessary and adds a significant amount of complication. Why not just add a suffix before passing the name into TraceThread() ?

**Comment by theuni** [comment]
From IRC:
> \<cfields\> jamesob: noo!
\<cfields\> jamesob: I hope you don't think I'm being NIH about that PR, that wasn't my intention at all. Coding up an alternative myself helps me to understand the challenges better, it wasn't meant to replace your work.

**Pull request closed without merging**